File: frontend/src/shared/schemas/personnelWork.ts
import { z } from "zod";

// schema ของผลงานครู (รายเทอม) + ไฟล์แนบ — ต้องตรงกับ backend
// (service.PersonnelWorkDTO / service.WorkFileDTO)

// ประเภทไฟล์แนบ — ตรงกับ backend domain (image|document|certificate)
export const workFileTypeSchema = z.enum(["image", "document", "certificate"]);
export type WorkFileType = z.infer<typeof workFileTypeSchema>;

export const workFileTypeLabel: Record<WorkFileType, string> = {
  image: "รูปภาพ",
  document: "เอกสาร",
  certificate: "เกียรติบัตร",
};

// --- ผลงาน ---
export const personnelWorkSchema = z.object({
  id: z.string(),
  title: z.string(),
  description: z.string(),
  work_date: z.string(),
  file_count: z.number(),
  created_at: z.string(),
});
export type PersonnelWorkRecord = z.infer<typeof personnelWorkSchema>;
export const personnelWorkListSchema = z.array(personnelWorkSchema);

export type WorkBody = {
  title: string;
  description: string;
  work_date: string;
};

// --- ไฟล์แนบ (url = signed URL หมดอายุ ตาม PDPA) ---
export const workFileSchema = z.object({
  id: z.string(),
  file_type: z.string(),
  original_name: z.string(),
  content_type: z.string(),
  size_bytes: z.number(),
  url: z.string(),
  created_at: z.string(),
});
export type WorkFileRecord = z.infer<typeof workFileSchema>;
export const workFileListSchema = z.array(workFileSchema);

File: backend/internal/domain/personnel_work.go
package domain

import "time"

// ประเภทไฟล์แนบของผลงานครู
const (
	WorkFileImage       = "image"
	WorkFileDocument    = "document"
	WorkFileCertificate = "certificate"
)

// ValidWorkFileType ตรวจว่าเป็นประเภทไฟล์ที่อนุญาต
func ValidWorkFileType(t string) bool {
	switch t {
	case WorkFileImage, WorkFileDocument, WorkFileCertificate:
		return true
	default:
		return false
	}
}

// PersonnelWork คือผลงานครู — ผูกเทอม (semester_id) ตาม CLAUDE.md ข้อ 4.3
type PersonnelWork struct {
	ID          string
	SchoolID    string
	SemesterID  string
	PersonnelID string
	Title       string
	Description string
	WorkDate    *time.Time
	FileCount   int // จำนวนไฟล์แนบ (เติมตอน list)
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

// NewPersonnelWork payload สร้างผลงาน
type NewPersonnelWork struct {
	Title       string
	Description string
	WorkDate    *time.Time
}

// UpdatePersonnelWork payload แก้ไขผลงาน
type UpdatePersonnelWork struct {
	Title       string
	Description string
	WorkDate    *time.Time
}

// PersonnelWorkFile คือไฟล์แนบของผลงาน — เข้าถึงผ่าน signed URL เท่านั้น (PDPA)
type PersonnelWorkFile struct {
	ID              string
	SchoolID        string
	PersonnelWorkID string
	FileType        string
	StoragePath     string
	OriginalName    string
	ContentType     string
	SizeBytes       int64
	CreatedAt       time.Time
}

// NewPersonnelWorkFile payload บันทึก metadata ไฟล์แนบ (หลังอัปโหลดขึ้น storage แล้ว)
type NewPersonnelWorkFile struct {
	FileType     string
	StoragePath  string
	OriginalName string
	ContentType  string
	SizeBytes    int64
}
